nipype/pipeline/engine/test_newengine/test_groupby.py

```diff
@@ -0,0 +1,163 @@
+from supernode import Node # or whatever
+
+import numpy as np
+
+import pytest, pdb
+from test_single_node import fun1, fun2, fun3
+from test_multi_nodes import fun4
+
+def fun2a(a, **dict):
+    pow = np.arange(3)
+    return a**pow
+
+
+@pytest.mark.parametrize("inputs_dict, expected_output", [ 
+        ({"a": [3, 4, 5]}, [(["a=3"], ([], 9)), (["a=4"], ([], 16)), (["a=5"], ([], 25))]), # lists are empty since there are no more inputs related to the values 
+        # do we want to allow 2D inputs a, when mapper="a"?
+        #({"a": [[3, 4, 5], [6, 7, 8]]}, [([3, 4, 5], [9, 16, 25]), ([6, 7, 8], [36, 49, 64])]),
+        ({"a": np.array([3, 4, 5])}, 
+         [(["a=3"], ([], 9)), (["a=4"], ([], 16)), (["a=5"], ([], 25))]),
+        ])
+def test_singlenode_groupbyind_1(inputs_dict, expected_output):
+    # for now i'm doing only groupByIndex, should I keep the original reducer name?
+    N  = Node(inputs=inuts_dict, mapper="a", groupByInd="a", interface=fun1)
+    N.run()
+    
+    # should we have both N.output and N.output_grouped/reduced ??
+    for (i, out) in enumerate(N.output["out"]):
+        assert out[0] == expected_output[i][0] 
+        assert (out[1][1] == expected_output[i][1][1]).all() #still assuming that values will be returned as arrays 
+
+    
+@pytest.mark.parametrize("inputs_dict, expected_output", [
+        ({"a":[3, 1, 8], "b":[0, 1, 2]}, 
+         [(["a=3"], (["b=0"], 0)), (["a=1"], (["b=1"], 1)), (["a=8"], (["b=2"], 16))]), # since there is a dot product, after grouping over "a" we have still only single values (but they are connected to values of "b") 
+        ])
+def test_single_groupbyind_2(inputs_dict, expected_output):
+    N = Node(interface=fun3, mapper=('a','b'), groupByInd="a")
```

Comment:
@effigies - very sorry for late answer, i completely missed your review 2 weeks ago... you're right that `groupyBy*` replacing temporary `reducer` (I think we didn't decide on the final name). I was suggesting `groupBy` notation, since in our examples we're not "reducing" anything but "grouping" output based on some criteria. 
I was concentrating mostly on "GroupByInd", since when you think about NumPy arrays it can be coded easier and I believed that this is the operation that you usually need. Generally speaking, if the input state contains only unique values, there is only one option. But if your input state looks like `a = [2, 3, 4, 2]` then `GroupByInd` gives four elements and `GroupByValue` gives three elements (combining output for `a[0]` and `a[3]`)